tensorflow_quantum/datasets/spin_system.py

```diff
@@ -0,0 +1,309 @@
+                exact diagonalization.
+            - `gs_energy`: Numpy `float` ground state energy from exact
+                diagonalization.
+            - `res_energy`: Python `float` residual between the circuit energy
+                and the exact energy from exact diagonalization.
+            - `fidelity`: Python `float` overlap between the circuit state
+                and the exact ground state from exact diagonalization.
+            -.`data_path`: Python `str` location of the data set.
+    """
+
+    supported_n = [4, 8, 12, 16]
+    supported_bc = ['closed']
+
+    if not all([isinstance(q, cirq.GridQubit) for q in qubits]):
+        raise TypeError("qubits must be a list of cirq.GridQubit objects.")
+
+    nspins = len(qubits)
+
+    if not all([q.col == 0 for q in qubits]):
+        raise ValueError(
+            'Must be list of cirq.GridQubit objects with shape ({},0)'.format(
+                nspins))
+
+    if not len(qubits) in supported_n:
+        raise ValueError('Supported number of spins are {}, received {}'.format(
+            supported_n, len(qubits)))
+
+    if boundary_condition not in supported_bc:
+        raise ValueError(
+            'Supported boundary conditions are {}, received {}'.format(
+                supported_bc, boundary_condition))
+
+    boundary_condition = boundary_condition
+    name = 'TFI_' + boundary_condition
+
+    def get_hamiltonian(qbs, g):
+        paulisum = []
+        for j in range(0, nspins - 1, 1):
+            paulisum.append(-cirq.Z(qbs[j]) * cirq.Z(qbs[j + 1]))
+        if boundary_condition == 'closed':
+            paulisum.append(-cirq.Z(qbs[nspins - 1]) * cirq.Z(qbs[0]))
+        for j in range(0, nspins):
+            paulisum.append(-g * cirq.X(qbs[j]))
+        hamiltonian = cirq.PauliString(paulisum[0])
+        for p in paulisum[1:]:
+            hamiltonian += p
+        return cirq.PauliSum.from_pauli_strings(hamiltonian)
+
+    dataset = SpinSystemDataset(nspins, name, data_dir)
+
+    name_generator = unique_name()
+    symbols = [sympy.Symbol(next(name_generator)) for _ in range(nspins)]
+    symbol_names = np.array([s.name for s in symbols]).reshape(
+        (2, int(nspins / 2)))
+    symbols = np.array(symbols).reshape((2, int(nspins / 2)))
+
+    # Define the circuit
+    circuit = cirq.Circuit()
+    circuit.append(cirq.H.on_each(qubits))
+
+    for d in range(int(nspins / 2)):
+        for j in range(0, nspins - 1, 2):
+            circuit.append(
+                cirq.ZZ(qubits[j], qubits[j + 1])**(symbols[0, d] / np.pi))
+        for j in range(1, nspins - 1, 2):
+            circuit.append(
+                cirq.ZZ(qubits[j], qubits[j + 1])**(symbols[0, d] / np.pi))
+        if boundary_condition == 'closed':
+            circuit.append(
+                cirq.ZZ(qubits[nspins - 1], qubits[0])**(symbols[0, d] / np.pi))
+        for j in range(0, nspins):
+            circuit.append(cirq.rx(symbols[1, d])(qubits[j]))
```

Comment:
In this block it seems to be setting up a circuit with scalar multiples in the symbols only to be immediately resolved. We don't keep a reference copy of the unresolved circuit in additional info so as a power user I have no way of experimenting with the values of these parameters because I don't know what goes where.

Perhaps in additional info we should store an unresolved copy of the circuit along with cirq.ParamResolver that allow users to do this.

Reply:
Great idea